File: locus-api-android-sample/src/main/java/com/asamm/locus/api/sample/ActivityDashboard.java
package com.asamm.locus.api.sample;

import android.os.Bundle;
import android.widget.TextView;

import com.asamm.locus.api.sample.receivers.PeriodicUpdateReceiver;
import com.asamm.locus.api.sample.utils.SampleCalls;

import java.text.SimpleDateFormat;
import java.util.Date;

import androidx.fragment.app.FragmentActivity;
import locus.api.android.features.periodicUpdates.PeriodicUpdatesHandler;
import locus.api.android.features.periodicUpdates.UpdateContainer;
import locus.api.android.utils.LocusUtils;
import locus.api.android.utils.LocusUtils.LocusVersion;
import locus.api.android.utils.exceptions.RequiredVersionMissingException;
import locus.api.utils.Logger;

public class ActivityDashboard extends FragmentActivity {

	// tag for logger
	private static final String TAG = "ActivityDashboard";

	// text containers
	private TextView tvInfo;
	private TextView tv01;
	private TextView tv02;
	private TextView tv03;
	private TextView tv04;
	private TextView tv05;
	private TextView tv06;
	private TextView tv07;
	private TextView tv08;

	// handler for updates
	private PeriodicUpdatesHandler.OnUpdate updateHandler =
			new PeriodicUpdatesHandler.OnUpdate() {

				@Override
				public void onUpdate(LocusVersion locusVersion, UpdateContainer update) {
					handleUpdate(update);
				}

				@Override
				public void onIncorrectData() {}
			};

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_dashboard);

		// prepare references to views
		tvInfo = findViewById(R.id.textView_info);
		tv01 = findViewById(R.id.textView1);
		tv02 = findViewById(R.id.textView2);
		tv03 = findViewById(R.id.textView3);
		tv04 = findViewById(R.id.textView4);
		tv05 = findViewById(R.id.textView5);
		tv06 = findViewById(R.id.textView6);
		tv07 = findViewById(R.id.textView7);
		tv08 = findViewById(R.id.textView8);
	}
	
	@Override
	public void onStart() {
		super.onStart();
		
		// register update handler
		try {
			PeriodicUpdateReceiver.setOnUpdateListener(this, updateHandler);
		} catch (RequiredVersionMissingException e) {
			Logger.logE(TAG, "onStart()", e);
		}

		// set info text
		handleUpdate(null);
	}
	
	@Override
	public void onStop() {
		super.onStop();
		
		// clear reference to prevent memory leaks
		try {
			PeriodicUpdateReceiver.setOnUpdateListener(this, null);
		} catch (RequiredVersionMissingException e) {
			Logger.logE(TAG, "onStop()", e);
		}
	}

	/**
	 * Handle fresh data.
	 * @param data received data
	 */
	private void handleUpdate(UpdateContainer data) {
		// check if data exists
		if (data == null) {
			LocusVersion activeVersion = LocusUtils.getActiveVersion(this);

			// prepare text info
			StringBuilder sb = new StringBuilder();
			sb.append("UpdateContainer not valid\n\n");
			sb.append("- active version: ").
					append(activeVersion.getVersionName()).
					append(" | ").
					append(activeVersion.getVersionCode()).
					append("\n");
			sb.append("- Locus Map is running: ").
					append(SampleCalls.isRunning(this, activeVersion) ? "running" : "stopped").
					append("\n");
			sb.append("- periodic updates: ").
					append(SampleCalls.isPeriodicUpdateEnabled(this, activeVersion) ? "enabled" : "disabled").
					append("\n");

			// set text to field
			tvInfo.setText(sb);
			return;
		}
		
		// refresh content
		tvInfo.setText("Fresh data received at " +
				SimpleDateFormat.getTimeInstance().format(new Date()));
		tv01.setText(String.valueOf(data.getLocMyLocation().getLatitude()));
		tv02.setText(String.valueOf(data.getLocMyLocation().getLongitude()));
		tv03.setText(String.valueOf(data.getGpsSatsUsed()));
		tv04.setText(String.valueOf(data.getGpsSatsAll()));
		tv05.setText(String.valueOf(data.isMapVisible()));
		tv06.setText(String.valueOf(data.getLocMyLocation().getAccuracy()));
		tv07.setText(String.valueOf(data.getLocMyLocation().getBearing()));
		tv08.setText(String.valueOf(data.getLocMyLocation().getSpeed()));
	}
}

File: locus-api-android-sample/src/main/java/com/asamm/locus/api/sample/ActivityDashboard.kt
package com.asamm.locus.api.sample

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import com.asamm.locus.api.sample.utils.SampleCalls
import locus.api.android.ActionBasics
import locus.api.android.features.periodicUpdates.UpdateContainer
import locus.api.android.objects.LocusVersion
import locus.api.android.utils.LocusUtils
import locus.api.utils.Logger
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * Sample activity that display method, how to periodically fetch base data from Locus Map
 * and display it, in this case, for example as simple dashboard.
 */
class ActivityDashboard : FragmentActivity() {

    // refresh handler
    private val handler: Handler = Handler(Looper.getMainLooper())
    // refresh interval (in ms)
    private val refreshInterval = TimeUnit.SECONDS.toMillis(1)
    // refresh task itself
    private val refresh: (() -> Unit) = {
        refreshContent()
    }

    // text containers
    private lateinit var tvInfo: TextView
    private lateinit var tv01: TextView
    private lateinit var tv02: TextView
    private lateinit var tv03: TextView
    private lateinit var tv04: TextView
    private lateinit var tv05: TextView
    private lateinit var tv06: TextView
    private lateinit var tv07: TextView
    private lateinit var tv08: TextView

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_dashboard)

        // prepare references to views
        tvInfo = findViewById(R.id.textView_info)
        tv01 = findViewById(R.id.textView1)
        tv02 = findViewById(R.id.textView2)
        tv03 = findViewById(R.id.textView3)
        tv04 = findViewById(R.id.textView4)
        tv05 = findViewById(R.id.textView5)
        tv06 = findViewById(R.id.textView6)
        tv07 = findViewById(R.id.textView7)
        tv08 = findViewById(R.id.textView8)
    }

    public override fun onStart() {
        super.onStart()

        // start refresh
        handler.post(refresh)
    }

    public override fun onStop() {
        super.onStop()

        // stop refresh flow
        handler.removeCallbacks(refresh)
    }

    /**
     * Perform refresh and request another update after defined interval.
     */
    private fun refreshContent() {
        try {
            Thread {
                LocusUtils.getActiveVersion(this)?.let { lv ->
                    ActionBasics.getUpdateContainer(this, lv)?.let { uc ->
                        handleUpdate(lv, uc)
                    } ?: {
                        handleUpdate(lv, null)
                        Logger.logW(TAG, "refreshContent(), " +
                                "unable to obtain `UpdateContainer`")
                    }()
                } ?: {
                    handleUpdate(null, null)
                    Logger.logW(TAG, "refreshContent(), " +
                            "unable to obtain `ActiveVersion`")
                }()

            }.start()
        } finally {
            handler.postDelayed(refresh, refreshInterval)
        }
    }

    /**
     * Handle fresh data.
     *
     * @param lv current Locus version
     * @param uc received container
     */
    @SuppressLint("SetTextI18n")
    private fun handleUpdate(lv: LocusVersion?, uc: UpdateContainer?) {
        handler.post {
            // check if uc exists
            if (lv == null || uc == null) {
                // prepare text info
                val sb = StringBuilder()
                sb.append("UpdateContainer not valid\n\n")
                sb.append("- active version: ")
                        .append(lv?.versionName).append(" | ").append(lv?.versionCode)
                        .append("\n")
                sb.append("- Locus Map is running: ")
                        .append(if (lv?.let { SampleCalls.isRunning(this, it) } == true) "running" else "stopped")
                        .append("\n")
                sb.append("- periodic updates: ")
                        .append(if (lv?.let { SampleCalls.isPeriodicUpdateEnabled(this, it) } == true) "enabled" else "disabled")
                        .append("\n")

                // set text to field
                tvInfo.text = sb
            } else {
                // refresh content
                tvInfo.text = "Fresh uc received at ${SimpleDateFormat.getTimeInstance().format(Date())}\n" +
                        "App: ${lv.versionName}, battery:${uc.deviceBatteryValue}"
                tv01.text = uc.locMyLocation.latitude.toString()
                tv02.text = uc.locMyLocation.longitude.toString()
                tv03.text = uc.gpsSatsUsed.toString()
                tv04.text = uc.gpsSatsAll.toString()
                tv05.text = uc.isMapVisible.toString()
                tv06.text = uc.locMyLocation.accuracyHor.value.toString()
                tv07.text = uc.locMyLocation.bearing.value.toString()
                tv08.text = uc.locMyLocation.speed.value.toString()
            }
        }
    }

    companion object {

        // tag for logger
        private const val TAG = "ActivityDashboard"
    }
}
